File: go/redundant_xlang_test.go
package multipath

import (
	"bufio"
	"bytes"
	"context"
	"io"
	"net"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"testing"
	"time"
)

// TestXLangJavaServerGoClient is the true cross-language end-to-end: a real JVM server process (the
// Kotlin RedundantServer, run via EchoServerMain) and a Go client, with a per-link fault middlebox
// between them doing black-hole / one-directional / disconnect cuts. The Go client sends a payload,
// the Java server echoes every byte back over the same redundant stream, and the round-trip must be
// byte-exact — proving the two implementations interoperate on the wire under adverse networks.
//
// Skipped unless MP_JVM_CP is set to the JVM classpath (see testbed/redundant-xlang/run.sh). This
// keeps the default `go test` free of a Java dependency.
func TestXLangJavaServerGoClient(t *testing.T) {
	cp := os.Getenv("MP_JVM_CP")
	if cp == "" {
		t.Skip("set MP_JVM_CP to the JVM classpath to run the cross-language e2e (see testbed/redundant-xlang/run.sh)")
	}
	const n = 3

	// Start the Java echo server on an ephemeral port; it prints "LISTENING <port>" when ready.
	cmd := exec.Command("java", "-cp", cp, "app.microteams.multipath.redundant.EchoServerMain", "0", strconv.Itoa(n))
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		t.Fatalf("stdout pipe: %v", err)
	}
	cmd.Stderr = os.Stderr
	if err := cmd.Start(); err != nil {
		t.Fatalf("start java: %v", err)
	}
	defer func() { _ = cmd.Process.Kill(); _ = cmd.Wait() }()

	br := bufio.NewReader(stdout)
	var srvPort int
	{
		line := make(chan string, 1)
		go func() {
			s, _ := br.ReadString('\n')
			line <- s
		}()
		select {
		case s := <-line:
			s = strings.TrimSpace(s)
			if !strings.HasPrefix(s, "LISTENING ") {
				t.Fatalf("unexpected server line: %q", s)
			}
			srvPort, err = strconv.Atoi(strings.TrimPrefix(s, "LISTENING "))
			if err != nil {
				t.Fatalf("parse port: %v", err)
			}
		case <-time.After(30 * time.Second):
			t.Fatalf("java server did not report LISTENING")
		}
	}

	// One fault middlebox per link, each fronting the Java server.
	boxes := make([]*middlebox, n)
	for i := 0; i < n; i++ {
		boxes[i] = newMiddlebox(t, "127.0.0.1:"+strconv.Itoa(srvPort), int64(2000+i))
		defer boxes[i].close()
	}

	cli, err := DialRedundant(context.Background(), RedundantOptions{
		N: n, Window: 2 << 20,
		PingInterval: 20 * time.Millisecond, DeadAfter: 100 * time.Millisecond,
		AckInterval: 8 * time.Millisecond, ReconnectDelay: 5 * time.Millisecond, MaxDelay: 40 * time.Millisecond,
		Dial: func(ctx context.Context, i int) (io.ReadWriteCloser, error) {
			var d net.Dialer
			return d.DialContext(ctx, "tcp", boxes[i].addr())
		},
	})
	if err != nil {
		t.Fatalf("client dial: %v", err)
	}
	defer cli.Close()

	msg := randBytes(256 << 10)
	go func() { _, _ = cli.Write(msg) }()

	got := make([]byte, 0, len(msg))
	buf := make([]byte, 64*1024)
	deadline := time.Now().Add(60 * time.Second)
	for len(got) < len(msg) {
		if time.Now().After(deadline) {
			t.Fatalf("timeout: echoed %d/%d bytes under fault injection", len(got), len(msg))
		}
		m, rerr := cli.Read(buf)
		if m > 0 {
			got = append(got, buf[:m]...)
		}
		if rerr != nil && len(got) < len(msg) {
			t.Fatalf("read err at %d/%d: %v", len(got), len(msg), rerr)
		}
	}
	if !bytes.Equal(got, msg) {
		t.Fatalf("cross-language echo mismatch")
	}
}

File: ts/test/redundant.xlang.test.ts
import { spawn } from 'node:child_process';
import { createInterface } from 'node:readline';
import { describe, expect, it } from 'vitest';
import { RedundantStream } from '../src/redundant.js';

// Cross-language proof for the redundant layer: the browser RedundantStream (run here in Node, whose
// global WebSocket stands in for the browser's) against the real JVM RedundantServer echoing bytes,
// over WebSocket links. Skipped unless MP_JVM_CP is set (see testbed/run.sh).
const cp = process.env.MP_JVM_CP;
const maybe = cp ? it : it.skip;

async function startEcho(n: number): Promise<{ port: number; kill: () => void }> {
  const proc = spawn('java', [
    '-cp',
    cp!,
    'app.microteams.multipath.redundant.EchoServerMain',
    '0',
    String(n),
  ]);
  proc.stderr.pipe(process.stderr);
  const rl = createInterface({ input: proc.stdout });
  const port = await new Promise<number>((resolve, reject) => {
    const timer = setTimeout(() => reject(new Error('java echo did not report LISTENING')), 30000);
    rl.on('line', (line) => {
      const m = /^LISTENING (\d+)/.exec(line);
      if (m) {
        clearTimeout(timer);
        resolve(Number(m[1]));
      }
    });
  });
  return { port, kill: () => proc.kill() };
}

describe('redundant stream cross-language', () => {
  maybe(
    'echoes a payload through a JVM redundant server over WebSocket links',
    async () => {
      const n = 3;
      const echo = await startEcho(n);
      try {
        const urls = Array.from({ length: n }, () => `ws://127.0.0.1:${echo.port}/mt/link`);
        const chunks: Uint8Array[] = [];
        let received = 0;
        let done!: () => void;
        const payload = new Uint8Array(64 * 1024).map((_, i) => (i * 31 + 7) & 0xff);
        const complete = new Promise<void>((resolve) => (done = resolve));

        const stream = new RedundantStream({
          urls,
          pingIntervalMs: 20,
          deadAfterMs: 2000,
          ackIntervalMs: 8,
          reconnectDelayMs: 20,
          maxDelayMs: 200,
        });
        stream.onDeliver = (bytes) => {
          chunks.push(bytes.slice());
          received += bytes.length;
          if (received >= payload.length) done();
        };
        await stream.dial();
        await stream.write(payload);
        await complete;
        stream.close();

        const got = new Uint8Array(received);
        let off = 0;
        for (const c of chunks) {
          got.set(c, off);
          off += c.length;
        }
        expect(got.subarray(0, payload.length)).toEqual(payload);
      } finally {
        echo.kill();
      }
    },
    40000,
  );
});
